### PR TITLE
fix: allow placeholders to be substited when coercible

### LINKED ISSUE #8976: Placeholders cannot be substituted with values when the values type is coercible but not equivalent
### Describe the bug

Currently when substituting placeholders with `LogicalPlan::with_param_values`, the types of the `ScalarValue`s used in the `ParamValues` collection must match exactly with the `Expr::Placeholder` expression being substituted, even when the `ScalarValue` can be coerced to the type of the `Expr::Placeholder` (e.g. `UInt64` to `Int64`). This is a bug caused by an eager check in the `ParamValues` code. See linked PR for more context.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kallisti-dev**
## Which issue does this PR close?

Closes #8976.

## Rationale for this change

It should be possible for the types of parameter values and placeholders to be different, as long as the `ScalarValue` is coercible to the `Expr::Placeholder`s data type.

## What changes are included in this PR?

Removal of the eager `==` check on data types, which allows type coercion to happen later in planning.

## Are there any user-facing changes?

The `data_type` argument to `ParamValues::get_placeholders_with_values` is currently unused. If we decide to remove it that would be a user-facing change to the API. 

Alternatively, instead of deleting the eager check and leaving the `data_type` parameter unused, we could improve the eager check with code borrowed from [the type coercion module](https://github.com/kallisti-dev/arrow-datafusion/blob/81f8078e49d1b9d877967de18960e993e5b3ee6a/datafusion/expr/src/type_coercion/functions.rs#L226-L234)  but I'm not sure of the best way to bridge the crate gap there (`datafusion-common` doesn't depend on `datafusion-expr`)

### CODE REVIEW

**Comment by kallisti-dev** [comment]
I think if `ParamValues` were moved to `datafusion-expr` that would solve the dependency loop here. Then it could call `can_coerce_from` directly.

**Comment by kallisti-dev** [comment]
Actually I think [this code](https://github.com/kallisti-dev/arrow-datafusion/blob/81f8078e49d1b9d877967de18960e993e5b3ee6a/datafusion/optimizer/src/analyzer/type_coercion.rs#L446-L461)  from the type coercion optimization rule is the logic we would want to use here

**Review by alamb — COMMENTED** [review]
Thank you @kallisti-dev  -- this is an important problem to solve

The code in this PR looks ok to me, but needs a test to make sure the problem is solved and to ensure we don't introduce regressions in the future

It would be amazing if you could take a shot at writing some tests, but if you don't have time I can do so as well.

Let me know if this does't make sense

# Test Outline
I think the test should probably look something like what we want to do in IOx (which I think is pretty standard). Maybe we can add a test to the `core_integration` tests in  `datafusion/core/tests/sql/parameters.rs` that does something like:

## Step 1: User provides a SQL with placeholder
for example, if `id` is an `i64`
```sql
SELECT * FROM t1 WHRERE id = $1
```

## DataFusion makes a `LogicalPlan` (possbly wrapped in a DataFrame)
```rust
let plan = ctx.sql(sql_string);
```

## User provides the parameters
(note in this case the user may provide an `i32` for `id` which should be automatically coerced to i64)

```rust
let plan = plan
  .with_param_values(vec![
    ("id", ScalarValue::from(3i32)),
  ])
```

# Tests to write;
1. Simply positive test (described above)
2. Negative tests (when the parameter value can not be coerced -- e.g. pass a string that is not a valid number)
3. Use both `PREPARE` and normal `SELECT`

**Comment by kallisti-dev** [comment]
@alamb  Added new tests to cover those cases. happy path for named params and prepared statements already existed so I put my new tests next to those.

However I noticed one test that should pass is not. It looks the string "not a number" is being successfully coerced to Int32

File: datafusion/core/tests/sql/select.rs [hunk 1]
```diff
@@ -572,6 +652,56 @@ async fn test_named_query_parameters() -> Result<()> {
...
+        ("unsigned", Arc::new(unsigned_ints) as ArrayRef),
+    ])?;
+    ctx.register_batch("test", batch)?;
+    let results = ctx
+        .sql("SELECT signed, unsigned FROM test WHERE signed = $text")
+        .await?
+        .with_param_values(vec![("text", ScalarValue::from("not a number"))])?
+        .collect()
+        .await;
+    assert_eq!(results.unwrap_err().strip_backtrace(), "");
```
Comment (kallisti-dev) [inline datafusion/core/tests/sql/select.rs:701]: this test is failing with

```
---- sql::select::test_parameter_invalid_types stdout ----
thread 'sql::select::test_parameter_invalid_types' panicked at datafusion/core/tests/sql/select.rs:701:24:
called `Result::unwrap_err()` on an `Ok` value: []
```

I would expect it to not be able to coerce the string

Comment (alamb) [reply datafusion/core/tests/sql/select.rs:704]: 🤔  I am also surprised this works, but apparently strings can be coerced to numbers (`null` in this case):

```
❯ create table test (signed int) as values (-1), (0), (1);
0 rows in set. Query took 0.005 seconds.

❯ select * from test;
+--------+
| signed |
+--------+
| -1     |
| 0      |
| 1      |
+--------+
3 rows in set. Query took 0.001 seconds.

❯ select * from test where signed = 'not a number';
0 rows in set. Query took 0.003 seconds.
```

Maybe you can try something like this:
```sql
❯ create table test3 as values ([1,2,3]);
0 rows in set. Query took 0.002 seconds.

❯ select * from test3;
+-----------+
| column1   |
+-----------+
| [1, 2, 3] |
+-----------+
1 row in set. Query took 0.001 seconds.

❯ select * from test3 where column1 = 4;
Arrow error: Invalid argument error: Invalid comparison operation: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) == List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
```

Here is an example of creating a list array: https://docs.rs/arrow/latest/arrow/array/type.ListArray.html#method.from_iter_primitive

**Review by alamb — DISMISSED** [review]
Looking very nice @kallisti-dev  👍 

I have a test suggestion on how to trigger an error but I think this PR could also be merged as is (if CI passes) and we could improve the tests as a follow on

File: datafusion/core/tests/sql/select.rs (hunk 1 above)
Comment (kallisti-dev) [reply datafusion/core/tests/sql/select.rs]: I've updated the test as described

**Comment by alamb** [comment]
> @alamb I've updated the failing test with the suggestion you made to try coercing integer to a list. it should pass now.

Looks like there is still a clippy failure @kallisti-dev :  https://github.com/apache/arrow-datafusion/actions/runs/7644208478/job/20828392768?pr=8977

**Review by alamb — APPROVED** [review]
Thank you @kallisti-dev  

I took the liberty of pushing a fix for clippy to your branch and I plan to merge this when CI passes

**Comment by alamb** [comment]
Thanks again @kallisti-dev

**Pull request merged**